cpp/tests/utilities/column_utilities.cu

```diff
@@ -536,28 +527,49 @@ struct column_comparator_impl {
                   size_type fp_ulps,
                   int depth)
   {
-    auto d_lhs = cudf::table_device_view::create(table_view{{lhs}});
-    auto d_rhs = cudf::table_device_view::create(table_view{{rhs}});
-
     auto d_lhs_row_indices = cudf::column_device_view::create(lhs_row_indices);
     auto d_rhs_row_indices = cudf::column_device_view::create(rhs_row_indices);
 
-    using ComparatorType = std::conditional_t<check_exact_equality,
-                                              corresponding_rows_unequal,
-                                              corresponding_rows_not_equivalent>;
+    auto d_lhs = cudf::column_device_view::create(lhs);
+    auto d_rhs = cudf::column_device_view::create(rhs);
 
     auto differences = rmm::device_uvector<int>(
       lhs.size(), cudf::get_default_stream());  // worst case: everything different
     auto input_iter = thrust::make_counting_iterator(0);
-    auto diff_iter  = thrust::copy_if(
-      rmm::exec_policy(cudf::get_default_stream()),
-      input_iter,
-      input_iter + lhs_row_indices.size(),
-      differences.begin(),
-      ComparatorType(*d_lhs, *d_rhs, *d_lhs_row_indices, *d_rhs_row_indices, fp_ulps));
 
-    differences.resize(thrust::distance(differences.begin(), diff_iter),
-                       cudf::get_default_stream());  // shrink back down
+    auto const comparator_helper = [&](auto const device_comparator) {
+      using ComparatorType =
+        std::conditional_t<check_exact_equality,
+                           corresponding_rows_unequal<decltype(device_comparator)>,
+                           corresponding_rows_not_equivalent<decltype(device_comparator)>>;
+
+      auto diff_iter = thrust::copy_if(
+        rmm::exec_policy(cudf::get_default_stream()),
+        input_iter,
+        input_iter + lhs_row_indices.size(),
+        differences.begin(),
+        ComparatorType(
+          *d_lhs_row_indices, *d_rhs_row_indices, fp_ulps, device_comparator, *d_lhs, *d_rhs));
+
+      differences.resize(thrust::distance(differences.begin(), diff_iter),
+                         cudf::get_default_stream());  // shrink back down
+    };
+
+    if constexpr (check_exact_equality) {
+      auto lhs_tview = table_view{{lhs}};
+      auto rhs_tview = table_view{{rhs}};
+
+      auto const comparator = cudf::experimental::row::equality::two_table_comparator{
+        lhs_tview, rhs_tview, cudf::get_default_stream()};
+      auto const has_nulls = cudf::has_nested_nulls(lhs_tview) or cudf::has_nested_nulls(rhs_tview);
+
+      auto const device_comparator = comparator.equal_to<false>(cudf::nullate::DYNAMIC{has_nulls});
+      comparator_helper(device_comparator);
+    } else {
+      // equivalence can be checked between column of different types,
+      // but the new comparator does not support that
+      comparator_helper(int{0});
+    }
```

Comment:
> equivalence can be checked between column of different types

We don't check different types. `check_exact_equality` is used for checking column "equal" or "equivalent".

Reply:
```suggestion
      auto lhs_tview = table_view{{lhs}};
      auto rhs_tview = table_view{{rhs}};

      auto const comparator = cudf::experimental::row::equality::two_table_comparator{
        lhs_tview, rhs_tview, cudf::get_default_stream()};
      auto const has_nulls = cudf::has_nested_nulls(lhs_tview) or cudf::has_nested_nulls(rhs_tview);

      auto const device_comparator = comparator.equal_to<false>(cudf::nullate::DYNAMIC{has_nulls});
      comparator_helper(device_comparator);
```

Reply:
We don't need to have such `comparator_helper`.

Reply:
Please look at the failures in CI https://github.com/rapidsai/cudf/actions/runs/4440315944/jobs/7794655639 in commit [d094524](https://github.com/rapidsai/cudf/pull/12777/commits/d094524d91abda6ee3eac65abd093a4d86f5fca7). `GetStructValueTest` fails because the comparator complains "cannot compare columns with different types"

Reply:
That test is probably incorrect:
```
cudf::test::dictionary_column_wrapper<TypeParam, uint32_t> f3{42, 42, 24};

cudf::test::dictionary_column_wrapper<int32_t, TypeParam> ef3{24};
```

When `TypeParam==short`, we have `f3` and `ef3` are of different types. Let's try to fix it. The previous row comparator doesn't check input types so it didn't throw.

Reply:
Isn't that the point of these tests? It's called `TYPED_TEST(StructGetValueTestTyped, mixed_types_valid)
` so I think it is checking mixed types

Reply:
Mixed types mean mixing these:
```
  cudf::test::fixed_width_column_wrapper<TypeParam> f1{1, 2, 3};
  cudf::test::strings_column_wrapper f2{"aa", "bbb", "c"};
  cudf::test::dictionary_column_wrapper<TypeParam, uint32_t> f3{42, 42, 24};
  LCW f4{LCW{8, 8, 8}, LCW{9, 9}, LCW{10}};
```